File: src/IsbInterop/ContextFactory.vb
Imports System.Data.SqlClient

''' <summary>
''' Фабрика контекста.
''' </summary>
Public NotInheritable Class ContextFactory

#Region "Константы"

  ''' <summary>
  ''' Формат строки подключения к DIRECTUM.
  ''' </summary>
  Private Const DirectumConnectionStringFormat As String = "ServerName={0};DBName={1};{2}"

  ''' <summary>
  ''' Часть строки подключения к DIRECTUM при windows-аутентификации.
  ''' </summary>
  Private Const DirectumOSAuthenticationPartString As String = "AuthType=OS"

  ''' <summary>
  ''' Часть строки форматирования для подключения к DIRECTUM при парольной аутентификации.
  ''' </summary>
  Private Const DirectumPasswordAuthenticationPartStringFormat As String = "UserName={0};Password={1}"

#End Region

  ''' <summary>
  ''' Создает контекст.
  ''' </summary>
  ''' <param name="connectionString">Строка подключения.</param>
  ''' <param name="storeInCache">Признак необходимости добавления информации о соединении в кэш.</param>
  ''' <returns>Контекст.</returns>
  Public Shared Function CreateContext(connectionString As String, Optional storeInCache As Boolean = True) As Context
    Dim connectionParams = GetConnectionParams(connectionString)
    Dim rcwApp = IsbApplicationManager.Instance.GetRcwApplication(connectionParams, storeInCache)

    Return New Context(rcwApp, connectionParams, storeInCache)
  End Function

  ''' <summary>
  ''' Создает контекст.
  ''' </summary>
  ''' <param name="storeInCache">Признак необходимости добавления информации о соединении в кэш.</param>
  ''' <returns>Контекст.</returns>
  Public Shared Function CreateContext(Optional storeInCache As Boolean = True) As Context
    Dim connectionString = ConfigurationUtils.GetConnectionString()
    Dim connectionParams = GetConnectionParams(connectionString)
    Dim rcwApp = IsbApplicationManager.Instance.GetRcwApplication(connectionParams, storeInCache)

    Return New Context(rcwApp, connectionParams, storeInCache)
  End Function

  ''' <summary>
  ''' Получает параметры подключения из строки подключения.
  ''' </summary>
  ''' <param name="connectionString">Строка подключения.</param>
  ''' <returns>Параметры подключения.</returns>
  Private Shared Function GetConnectionParams(connectionString As String) As String
    If connectionString Is Nothing Then
      Throw New FatalIsbInteropException(My.Resources.Resources.UnableToGetDBConnectionParams)
    End If

    Dim builder = New SqlConnectionStringBuilder(connectionString)
    Dim user = builder.UserID

    Dim authenticationPart As String

    If builder.IntegratedSecurity Then
      authenticationPart = DirectumOSAuthenticationPartString
    Else
      authenticationPart = String.Format(DirectumPasswordAuthenticationPartStringFormat, user, builder.Password)
    End If

    Dim server = builder.DataSource
    Dim database = builder.InitialCatalog

    Dim isbConnectionParams = String.Format(DirectumConnectionStringFormat, server, database, authenticationPart)
    Return isbConnectionParams
  End Function
End Class

File: src/IsbInterop.CSharp/ContextFactory.cs
using IsbInterop.Properties;
using System.Data.SqlClient;

namespace IsbInterop
{
  /// <summary>
  /// Фабрика контекста.
  /// </summary>
  public static class ContextFactory
  {
    #region Константы

    /// <summary>
    /// Формат строки подключения к DIRECTUM.
    /// </summary>
    private const string DirectumConnectionStringFormat = @"ServerName={0};DBName={1};{2}";

    /// <summary>
    /// Часть строки подключения к DIRECTUM при windows-аутентификации.
    /// </summary>
    private const string DirectumOSAuthenticationPartString = "AuthType=OS";

    /// <summary>
    /// Часть строки форматирования для подключения к DIRECTUM при парольной аутентификации.
    /// </summary>
    private const string DirectumPasswordAuthenticationPartStringFormat = @"UserName={0};Password={1}";

    #endregion

    /// <summary>
    /// Создает контекст.
    /// </summary>
    /// <param name="connectionString">Строка подключения.</param>
    /// <param name="storeInCache">Признак необходимости добавления информации о соединении в кэш.</param>
    /// <returns>Контекст.</returns>
    public static Context CreateContext(string connectionString, bool storeInCache = true)
    {
      var connectionParams = GetConnectionParams(connectionString);
      var rcwApp = IsbApplicationManager.Instance.GetRcwApplication(connectionParams, storeInCache);

      return new Context(rcwApp, connectionParams, storeInCache);
    }

    /// <summary>
    /// Создает контекст.
    /// </summary>
    /// <param name="storeInCache">Признак необходимости добавления информации о соединении в кэш.</param>
    /// <returns>Контекст.</returns>
    public static Context CreateContext(bool storeInCache = true)
    {
      var connectionString = ConfigurationUtils.GetConnectionString();
      var connectionParams = GetConnectionParams(connectionString);
      var rcwApp = IsbApplicationManager.Instance.GetRcwApplication(connectionParams, storeInCache);

      return new Context(rcwApp, connectionParams, storeInCache);
    }

    /// <summary>
    /// Получает параметры подключения из строки подключения.
    /// </summary>
    /// <param name="connectionString">Строка подключения.</param>
    /// <returns>Параметры подключения.</returns>
    private static string GetConnectionParams(string connectionString)
    {
      if (connectionString == null)
        throw new FatalIsbInteropException(Resources.UnableToGetDBConnectionParams);

      var builder = new SqlConnectionStringBuilder(connectionString);
      var user = builder.UserID;

      string authenticationPart;

      if (builder.IntegratedSecurity)
        authenticationPart = DirectumOSAuthenticationPartString;
      else
      {
        authenticationPart = string.Format(DirectumPasswordAuthenticationPartStringFormat,
          user, builder.Password);
      }

      var server = builder.DataSource;
      var database = builder.InitialCatalog;

      var isbConnectionParams = string.Format(DirectumConnectionStringFormat, server, database, authenticationPart);
      return isbConnectionParams;
    }
  }
}
